### PR TITLE
Crusade Haste mod and T20 4P

### PR DESCRIPTION
**Pull request by CollCrom**

### CODE REVIEW

File: src/Parser/RetributionPaladin/Modules/Items/Tier20_2set.js [hunk 1]
```diff
@@ -48,17 +48,16 @@ class Tier20_2set extends Module {
...
-	// 				.icon(SPELLS.RET_PALADIN_T20_2SET_BONUS_BUFF.id)
-	// 				.actual(`${formatPercentage(actual)}% uptime`)
-	// 				.recommended(`${formatPercentage(recommended)}% is recommended`)
-	// 				.regular(recommended).major(recommended - 0.05);
-	// 		});
-	// }
+	suggestions(when) {
+	 	when(this.percentUptime).isLessThan(.95)
+	 		.addSuggestion((suggest, actual, recommended) => {
+ 				return suggest(`Your Tier 20 2pc uptime of ${formatPercentage(actual)}% is below 95%, try to use judgement as much as possible`)
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Items/Tier20_2set.js]: When reviewing a log would you really look at the tier 20 2 set uptime and tell someone this? I imagine you'd look if someone has the 2PT20 and if so you would look at his Judgment casts, and if this is too low tell him to cast Judgment more. In that case I would recommend moving this suggestion to Judgment in the CastEfficiency module; you should be able to set it as an `extraSuggestion` with `isActive` toggling between default and 2PT20 equipped checks.

File: src/Parser/RetributionPaladin/Modules/Items/Tier20_4set.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	holyPowerGained = 0;
+	builderId = SPELLS.BLADE_OF_JUSTICE.id;
+
+	on_byPlayer_cast(event) {
+		if(this.combatants.selected.hasBuff(SPELLS.RET_PALADIN_T20_4SET_BONUS_BUFF.id) && (event.ability.guid === SPELLS.BLADE_OF_JUSTICE.id || event.ability.guid === SPELLS.DIVINE_HAMMER_TALENT.id)){
+			//The Tier bonus turns our 2 HP builders into 3 HP builders
+			this.holyPowerGained += 1;
+		}
+		if(this.combatants.selected.hasTalent(SPELLS.DIVINE_HAMMER_TALENT.id)){
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Items/Tier20_4set.js]: This should be in the `on_initialized` event handler.

Comment (MartijnHols) [reply src/Parser/RetributionPaladin/Modules/Items/Tier20_4set.js]: Actually seeing as the only place this value is used is in `item()`, you can also put it in the `item()` method, making it a local variable.

File: src/Parser/RetributionPaladin/Modules/Features/CastEfficiency.js [hunk 3]
```diff
@@ -60,6 +60,15 @@ class CastEfficiency extends CoreCastEfficiency {
...
       getCooldown: haste => 12 / (1 + haste),
     },
+    //This is the judgment CE with t20
+    {
+      spell: SPELLS.JUDGMENT_CAST,
+      category: CastEfficiency.SPELL_CATEGORIES.ROTATIONAL,
+      getCooldown: haste => 12 / (1 + haste),
+      isActive: combatant => combatant.hasBuff(SPELLS.RET_PALADIN_T20_2SET_BONUS_BUFF.id),
+      recommendedCastEfficiency: 0.95,
+      extraSuggestion: 'With tier 20 2 peice is is even more imporant to use Judgment on cooldown to keep up the buff',
```
Comment (rp4rk) [inline src/Parser/RetributionPaladin/Modules/Features/CastEfficiency.js]: Typo here

File: src/Parser/RetributionPaladin/Modules/Items/Tier20_4set.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+class Tier20_4set extends Module {
+	static dependencies = {
+		combatants: Combatants,
+	};
+
+	holyPowerGained = 0;
+
+	on_byPlayer_cast(event) {
+		if(this.combatants.selected.hasBuff(SPELLS.RET_PALADIN_T20_4SET_BONUS_BUFF.id) && (event.ability.guid === SPELLS.BLADE_OF_JUSTICE.id || event.ability.guid === SPELLS.DIVINE_HAMMER_TALENT.id)){
```
Comment (rp4rk) [inline src/Parser/RetributionPaladin/Modules/Items/Tier20_4set.js]: This is a huge `if` statement, can you abstract your logic for determining a valid cast into a method? Something like `benefitsFrom4Pc`.

**Comment by MartijnHols** [comment]
Thanks!

**Pull request merged**